### PR TITLE
NIFI-4434 Fixed recursive listing with a custom regex filter.

### PR DESCRIPTION
**Pull request by jtstorck**
Filter modes are now supported to perform listings based on directory and file names, file-names only, and full path.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by jtstorck** [comment]
[PR 2930](https://github.com/apache/nifi/pull/2930) was closed due to the branch in my fork being removed before adding the new filter-mode-based changes.  @bbende @ottobackwards, this PR implements the use cases discussed in the previous PR:
- filename only
- filename and directory name
- full path

**Comment by ottobackwards** [comment]
@jtstorck will review.
First quick thing is to ask if you have considered adding an additionDetails talking about why you would choose one strategy over another, maybe with simple examples?

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 1]
```diff
@@ -462,11 +523,15 @@ private String getPerms(final FsAction action) {
 
     private PathFilter createPathFilter(final ProcessContext context) {
         final Pattern filePattern = Pattern.compile(context.getProperty(FILE_FILTER).getValue());
```
Comment (ottobackwards) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: Does this need to support expression language?

Comment (jtstorck) [reply nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: @ottobackwards The FILE_FILTER property does not currently support expression language.  The processor could be updated to enable EL for the property, but that is outside the scope of this PR.

**Comment by jtstorck** [comment]
@ottobackwards Regarding documentation for the filter modes, descriptions have been created for the allowable values.  Do these descriptions not seem adequate for the functionality of each mode?

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java (hunk 1 above)
Comment (ottobackwards) [reply nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: fair enough

**Comment by ottobackwards** [comment]
@jtstorck I think that examples would help the user get the right one the first time.  Not a deal breaker though.

**Comment by jtstorck** [comment]
@ottobackwards @zenfenan I'll break out my rusty HTML skills and try to write up some extra documentation with examples/usecases.  Hopefully in an hour or two I'll have an update for the PR.  Thanks for the reviews!

**Comment by jtstorck** [comment]
@ottobackwards @zenfenan I added additional details to give some examples on how to use the filters.  Please let me know if you think more detail is needed.  I appreciate the review!

**Comment by zenfenan** [comment]
+1 Thanks @jtstorck. Merged to the master. Thanks @ottobackwards for reviewing.

**Pull request closed without merging**